src/interface/mod.rs

```diff
@@ -20,368 +20,608 @@
-#[serde(rename_all = "snake_case")]
-pub(crate) struct BaseInterface {
+#[derive(Deserialize, Debug, PartialEq, Clone)]
+#[serde(try_from = "InterfaceDef")]
+pub struct Interface {
     interface_name: String,
     version_major: i32,
     version_minor: i32,
     ownership: Ownership,
-    #[serde(skip_serializing_if = "Option::is_none")]
     description: Option<String>,
-    #[serde(skip_serializing_if = "Option::is_none")]
     doc: Option<String>,
+    inner: InterfaceIml,
 }
 
-#[doc(hidden)]
-#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
-pub struct DatastreamInterface {
-    #[serde(flatten)]
-    base: BaseInterface,
-    #[serde(default, skip_serializing_if = "is_default")]
-    aggregation: Aggregation,
-    mappings: Vec<DatastreamMapping>,
-}
+impl Interface {
+    /// Instantiate a new `Interface` from a file.
+    pub fn from_file(path: &Path) -> Result<Self, Error> {
+        let file = fs::read_to_string(path)?;
 
-#[doc(hidden)]
-#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
-pub struct PropertiesInterface {
-    #[serde(flatten)]
-    base: BaseInterface,
-    mappings: Vec<PropertiesMapping>,
-}
+        Self::from_str(&file)
+    }
 
-#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
-#[serde(rename_all = "snake_case")]
-pub(crate) enum InterfaceType {
-    Datastream,
-    Properties,
-}
+    /// Returns the interface name.
+    pub fn interface_name(&self) -> &str {
+        &self.interface_name
+    }
 
-#[derive(Serialize, Deserialize, Debug, PartialEq, Copy, Clone)]
-#[serde(rename_all = "snake_case")]
-pub(crate) enum Ownership {
-    Device,
-    Server,
-}
+    /// Returns the interface major version.
+    pub fn version_major(&self) -> i32 {
+        self.version_major
+    }
 
-#[derive(Serialize, Deserialize, Debug, PartialEq, Copy, Clone)]
-#[serde(rename_all = "snake_case")]
-pub(crate) enum Aggregation {
-    Individual,
-    Object,
-}
+    /// Returns the interface minor version.
+    pub fn version_minor(&self) -> i32 {
+        self.version_minor
+    }
 
-#[derive(Debug)]
-pub(crate) enum Mapping<'a> {
-    Datastream(&'a DatastreamMapping),
-    Properties(&'a PropertiesMapping),
-}
+    /// Returns the interface version.
+    fn version(&self) -> (i32, i32) {
+        (self.version_major, self.version_minor)
+    }
 
-#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
-pub(crate) struct BaseMapping {
-    endpoint: String,
-    #[serde(rename = "type")]
-    mapping_type: MappingType,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    description: Option<String>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    doc: Option<String>,
-}
+    /// Returns the interface type.
+    pub fn interface_type(&self) -> InterfaceType {
+        match &self.inner {
+            InterfaceIml::DatastreamIndividual(_) => InterfaceType::Datastream,
+            InterfaceIml::DatastreamObject(_) => InterfaceType::Datastream,
```

Comment:
Join the two statements.
```suggestion
            InterfaceIml::DatastreamIndividual(_) | 
            InterfaceIml::DatastreamObject(_) => InterfaceType::Datastream,
```

Reply:
Done